src/librustc_save_analysis/dump_visitor.rs

```diff
@@ -139,6 +145,137 @@ impl<'l, 'tcx: 'l, 'll, D: Dump + 'll> DumpVisitor<'l, 'tcx, 'll, D> {
+        let loans = mem::replace(&mut analysis_result.all_loans, vec![])
+            .into_iter()
+            .filter_map(|loan| self.process_loan(def_id, loan))
+            // Chain safe loans
+            .chain(mem::replace(&mut analysis_result.safe_loans, vec![])
+                .into_iter()
+                .filter_map(|safe_loan| self.process_safe_loan(safe_loan)))
+            .collect::<Vec<_>>();
+
+        let moves = mem::replace(&mut *analysis_result.move_data.moves.borrow_mut(), vec![])
+            .into_iter()
+            .map(|m| self.process_move(m, &analysis_result.move_data))
+            .collect::<Vec<_>>();
+
+        let scopes = {
+            let move_map: HashMap<Id, &MoveData> = moves.iter()
+                .fold(HashMap::new(), |mut acc, m| {
+                    match acc.entry(m.ref_id) {
+                        // We want the earliest occurring move
+                        Entry::Occupied(mut e) => if e.get().span.byte_start > m.span.byte_start { e.insert(m); },
+                        Entry::Vacant(e) => { e.insert(m); },
+                    }
+
+                    acc
+                });
+            mem::replace(&mut *analysis_result.move_data.var_assignments.borrow_mut(), vec![])
+                .into_iter()
+                .filter_map(|a| {
+                    let id = ::id_from_node_id(a.assignee_id, &self.save_ctxt);
+                    let mov = move_map.get(&id);
+                    self.process_assignment(a, mov, &analysis_result.move_data, &analysis_result.bccx)
+                })
+                .collect()
+        };
+
+        let span = analysis_result.span.map(|s| self.span_from_span(s));
+
+        let data = BorrowData {
+            ref_id: ::id_from_def_id(def_id),
+            scopes: scopes,
+            loans: loans,
+            moves: moves,
+            span: span,
+        };
+
+        self.dumper.dump_per_fn_borrow_data(data);
+    }
+
+    fn process_assignment(
+        &self,
+        assignment: Assignment,
+        mov: Option<&&MoveData>,
+        move_data: &rustc_borrowck::MoveData<'tcx>,
+        bccx: &BorrowckCtxt<'l, 'tcx>)
+    -> Option<Scope> {
+        move_data.path_loan_path(assignment.path)
+            .kill_scope(&bccx)
+            .span(&self.tcx.hir)
+            .map(|span| {
+                let expanded_span = match mov.map(|m| &m.span) {
+                    Some(move_span) => if move_span.byte_start < span.hi.0 {
+                        Span {
+                            lo: span.lo,
+                            hi: BytePos(move_span.byte_end),
+                            ctxt: NO_EXPANSION,
+                        }
+                    } else {
+                        span
+                    },
+                    None => span,
+                };
+
+                Scope {
+                    ref_id: ::id_from_node_id(assignment.assignee_id, &self.save_ctxt),
```

Comment:
I need to fix these. Working on the visualization integration, I ran into obvious problems here. `id_from_node_id` looks for local `DefId`s before converting to `rls-data`'s Id. The problem is that these `NodeId`s (and the ones below) are referring to Local nodes, which have no `DefId`. If the type of this object isn't in the same crate, these all get converted to NULL Ids. I didn't run into this earlier because I was testing with a single file crate (sorry about that).

@nrc does it make sense to create a new type of `Id` that's used only so these `BorrowData` types can reference each other?